### PR TITLE
planner, util/ranger:  apply PushDownNot to condition before pruning partition

### LINKED ISSUE #23608: wrong partition pruning cause the query result incorrect
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```sql
drop table if exists t;
create table t(a int) partition by hash (a) partitions 10;
insert into t values (1), (2), (12), (3), (11), (13);
select * from t where a not between 2 and 2;
explain select * from t where a not between 2 and 2;
```

### 2. What did you expect to see? (Required)
```sql
MySQL > select * from t where a not between 2 and 2;
+----+
| a  |
+----+
| 1  |
| 11 |
| 12 |
| 3  |
| 13 |
+----+

```
### 3. What did you see instead (Required)
```sql
mysql > select * from t where a not between 2 and 2;
+----+
| a  |
+----+
| 12 |
+----+

mysql > explain select * from t where a not between 2 and 2;
+---------------------+---------+-----------+-----------------------+--------------------------------------+
| id                  | estRows | task      | access object         | operator info                        |
+---------------------+---------+-----------+-----------------------+--------------------------------------+
| TableReader_9       | 1.33    | root      |                       | data:Selection_8                     |
| └─Selection_8       | 1.33    | cop[tikv] |                       | or(lt(test.t.a, 2), gt(test.t.a, 2)) |
|   └─TableFullScan_7 | 2.00    | cop[tikv] | table:t, partition:p2 | keep order:false, stats:pseudo       |
+---------------------+---------+-----------+-----------------------+--------------------------------------+


```

### 4. What is your TiDB version? (Required)
```sql
mysql > select tidb_version()\G
***************************[ 1. row ]***************************
tidb_version() | Release Version: v4.0.0-beta.2-2477-g149440854-dirty
Edition: Community
Git Commit Hash: 149440854e539ba5e674109f96af13cf9e34fe7f
Git Branch: master
UTC Build Time: 2021-03-26 13:48:42
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
```

### LINKED ISSUE #23617: the result of hash partition table with not expression in where clause is incorrect
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)
```
drop table if exists t;
create table t(a int) partition by hash (a) partitions 5;
insert into t values (0);
select * from t where not (a < -10 or a > 10);
select * from t where not (a < -10);
select * from t where not (a > 10);
```

### 2. What did you expect to see? (Required)
```
mysql> select * from t where not (a < -10 or a > 10);
+------+
| a    |
+------+
|    0 |
+------+
1 row in set (0.00 sec)

mysql> select * from t where not (a < -10);
+------+
| a    |
+------+
|    0 |
+------+
1 row in set (0.00 sec)

mysql> select * from t where not (a > 10);
+------+
| a    |
+------+
|    0 |
+------+
1 row in set (0.00 sec)
```

### 3. What did you see instead (Required)
```
mysql> select * from t where not (a < -10 or a > 10);
Empty set (0.00 sec)

mysql> select * from t where not (a < -10);
Empty set (0.00 sec)

mysql> select * from t where not (a > 10);
Empty set (0.00 sec)
```

### 4. What is your TiDB version? (Required)
```

mysql> select tidb_version()\G
*************************** 1. row ***************************
tidb_version(): Release Version: v4.0.0-beta.2-2479-g1e3921aff
Edition: Community
Git Commit Hash: 1e3921aff5f7f2bcc02a3d520244ba1db547d734
Git Branch: master
UTC Build Time: 2021-03-28 12:43:33
GoVersion: go1.13.15
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false
1 row in set (0.01 sec)
```

### 5. Root Cause:
In `func (r *builder) buildFromNot(expr *expression.ScalarFunction) []*point`, we don't handle the case when `expr.FuncName.L` is `LogicOr`, `GT`, `LT` and so on. To be on the safe side, for those unhandled cases, we need to return `getFullRange()` rather than `nil`.

### PR DESCRIPTION
**Pull request by xuyifangreeneyes**
### What problem does this PR solve?

Issue Number: close #23608, close #23617 

Problem Summary:
The result of hash partition table with not expression in where clause is incorrect.

### What is changed and how it works?

What's Changed & How it Works:
Apply PushDownNot to condition before pruning partition.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

- fix the wrong result of querying hash partition table with not expression in where clause

### CODE REVIEW

**Review by winoros — APPROVED** [review]
A better expression framework to not let us add `PushDownNot` by hand is indeed needed.

File: util/ranger/points.go [hunk 1]
```diff
@@ -638,8 +638,6 @@ func (r *builder) buildFromNot(expr *expression.ScalarFunction) []*point {
 		startPoint := &point{value: types.MinNotNullDatum(), start: true}
 		endPoint := &point{value: types.MaxValueDatum()}
 		return []*point{startPoint, endPoint}
-	case ast.LogicAnd:
```
Comment (eurekaka) [inline util/ranger/points.go:641]: Why do we remove this branch? we cannot guarantee that we have made all the necessary `PushDownNot` calls at least for now?

Comment (xuyifangreeneyes) [reply util/ranger/points.go:641]: https://github.com/pingcap/tidb/pull/22914#discussion_r602854476, this branch is not properly implemented. We need to implement `LogicOr`, `LT`, `GT` and so on. Shall we implement all these cases in this PR or just return `getFullRange()` for currently unimplemented cases?

Comment (eurekaka) [reply util/ranger/points.go:641]: OK, we can leave the optimizations to a separate PR, and focus on the correctness problem in this PR.

**Comment by xuyifangreeneyes** [comment]
PTAL @eurekaka

**Review by eurekaka — APPROVED** [review]
/lgtm

**Pull request merged**